### ISSUE TITLE
Roundup: more documentation/design questions

### ISSUE THREAD

**Issue by woodruffw** [sigstore/protobuf-specs#44]
I'm doing another one of these roundup issues 🙂 

## `common.PublicKeyIdentifier`

```protobuf
// PublicKeyIdentifier can be used to identify an (out of band) delivered
// key, to verify a signature.
message PublicKeyIdentifier {
        // Optional unauthenticated hint on which key to use.
        // The format of the hint must be agreed upon out of band by the
        // signer and the verifiers, and so is not subject to this
        // specification.
        // Example use-case is to specify the public key to use, from a
        // trusted key-ring.
        // Implementors are RECOMMENDED derive the value from the public
        // key as described in https://www.rfc-editor.org/rfc/rfc3280#section-4.2.1.1
        string hint = 1;
}
```

The guidance in RFC3280 Ss. 4.2.1.1 isn't great: it encourages identification with the SHA-1 of the ASN.1 `BIT STRING` for the `subjectPublicKeyInfo` (presumably DER encoded), with some modifications made to it.

Maybe we could recommend <https://www.rfc-editor.org/rfc/rfc6962#section-3.2> instead, since it's what we're already doing for key identification elsewhere due to CT? It's a lot simpler (just `SHA2-256(DER(subjectPublicKeyInfo))`), and uses a more modern digest.

## Verification messages

I might be missing how these are used/expected to be used, but some questions absent that knowledge:

1. Are these intended to be building blocks for verification policies, i.e. verifying specific OIDC claims/X.509v3 extensions once signature, etc. verification has succeeded? 
    1. If that's the intent, it might be worth thinking about regular expression support in `ObjectIdentifierValuePair`, similar to the current support in `SubjectAlternativeName`.
    2. Additionally, it might be nice to have building blocks for composing policies, similar to how we've designed the policy API in `sigstore-python` -- it would be nice to be able to compose messages that describe "all of"/"any of" and other set-language operations (like thresholds)
1. How does `Artifact` interact with a bundle's `MessageSignature.message_digest`? Is the expectation here that the verifier would collect the `Artifact` (either as direct bytes or from the specified URL) and confirm that it matches the `message_digest`? I think this is similar to the question I had in https://github.com/sigstore/protobuf-specs/issues/30, but I want to make sure I fully understand the moving pieces here 🙂 

cc @znewman01 @haydentherapper @kommendorkapten

**Comment by haydentherapper** [comment]
> Maybe we could recommend https://www.rfc-editor.org/rfc/rfc6962#section-3.2 instead, since it's what we're already doing for key identification elsewhere due to CT? It's a lot simpler (just SHA2-256(DER(subjectPublicKeyInfo))), and uses a more modern digest.

+1. 3820 is actually a little difficult to work with since most crypto libraries mark SHA-1 as deprecated.

**Comment by woodruffw** [comment]
Yep -- I'm wondering where that recommendation even came from, since 3820 is pretty old. Maybe this came from TUF somewhere? If so, TUF's own guidance for key IDs might need updating 🙂

**Comment by asraa** [comment]
>  If so, TUF's own guidance for key IDs might need updating slightly_smiling_face

TUF did make key IDs optional and removed the requirement for any kind of fixed algorithm for key id calculation (leaving just that it must be unique). Some implementations use sha256 like you mentioned.

I'm not exactly sure where the SHA-1 came from.

**Comment by haydentherapper** [comment]
Might be because we have to calculate subject key IDs in [sigstore/sigstore](https://github.com/sigstore/sigstore/blob/1399ee5c5de44ddc82d03e1fa85e997e85390229/pkg/cryptoutils/publickey.go#L84-L100), which are based on 3280

**Comment by kommendorkapten** [comment]
Recommending something that does not rely on SHA-1 is totally fine, as it is still a recommendation, the field itself is explicitly marked as optional. It may even contain references to KMS keys, or a specific version of a vault key as proposed in [this](https://github.com/sigstore/cosign/issues/1351#issuecomment-1285501313) comment by @znewman01 

> Are these intended to be building blocks for verification policies, i.e. verifying specific OIDC claims/X.509v3 extensions once signature, etc. verification has succeeded?

Yes, that is the intent. And I agree that supporting a regular expression is probably a good idea. We are just building out these messages now, and don't want to over-engineer before we have a clear use-case. There are most likely more use-cases than what I've heard of, but a regexp for the SAN and a few static oids are one use-case I'm aware of that e.g. the npm integration is going go use. If we have more concrete use-cases now I think we should try to update now of course, before we cut a release.

> How does Artifact interact with a bundle's MessageSignature.message_digest?

Yes, as you said it should be verified that the provided artifact indeed is the one referenced by the bundle (in the case of a DSSE envelope, the artifact is embedded into the envelope). 
Another scenario is in the case of using Ed25519 signatures, which must have access to the entire artifact's bytes to be able to verify the signature.

**Comment by woodruffw** [comment]
Thanks for clarifying!

I can do a quick PR to tweak the key ID guidance there.


> Yes, that is the intent. And I agree that supporting a regular expression is probably a good idea. We are just building out these messages now, and don't want to over-engineer before we have a clear use-case. There are most likely more use-cases than what I've heard of, but a regexp for the SAN and a few static oids are one use-case I'm aware of that e.g. the npm integration is going go use. If we have more concrete use-cases now I think we should try to update now of course, before we cut a release.

Makes sense to me! Something I've been thinking about: OIDs don't have great UX, so it might make sense for the Sigstore ecosystem to standardize on human-friendly names for the standard Fulcio OIDs (in <https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md>). 

For example, instead of something resembling this:

```yaml
1.3.6.1.4.1.57264.1.4: ci.yml
```

we'd have this:

```yaml
github-workflow-name: ci.yml
```

OTOH, since the bundles in question are mostly going to be machine generated, maybe this doesn't matter so much.

**Comment by kommendorkapten** [comment]
> For example, instead of something resembling this:

I think this can be done in each language client, protobuf does not IIRC support any const values. So e.g. in the client spec we can define some well-known OIDs that each client should recognize. But as there may be a scenario where a "any" oid is used, I think it's safer to keep them numerical. I also would think that a policy is write-once, read-many so not sure of how big burden it would actually be to work with numerical values?

**Comment by znewman01** [comment]
A few misc. points:

> Additionally, it might be nice to have building blocks for composing policies, similar to how we've designed the policy API in sigstore-python -- it would be nice to be able to compose messages that describe "all of"/"any of" and other set-language operations (like thresholds)

Soapbox: I think that eventually this is going to wind up reinventing a logic programming dialect 😄 (I don't necessarily mean that as a point of opposition, but I think that's the logical conclusion of "allow users to provide verification policies" (and I'm here for it).

> We are just building out these messages now, and don't want to over-engineer before we have a clear use-case.

+1, I think all of these enhancements are good ideas but would prefer to wait until there's real demand.

> So e.g. in the client spec we can define some well-known OIDs that each client should recognize. 

I don't *hate* the idea of putting these in protobuf-specs (though I don't know if it's even possible, as @kommendorkapten points out), just so clients in every language can automatically grab some well-known OIDs rather than having to inline the constants themselves. But overall +1 to keeping these in the client and not the bundle.

I added a TODO to the [client spec](https://docs.google.com/document/d/1kbhK2qyPPk8SLavHzYSDM8-Ueul9_oxIMVFuWMWKz0E/edit#) (join sigstore-dev@googlegroups.com for access).

--------------------------------------------------------------------------------------------------

What needs to be done before we can close this out? Maybe:

- [ ] add some more description to the Verification messages describing intended use
- [ ] add some docs on the Artifact <-> Message.message_digest relationship
- [ ] file a FR about regex support for things other than SAN in the Verification messages

**Comment by woodruffw** [comment]
I think this conversation ran stale and the main question (about key IDs) has been resolved, so closing!
